Kwicks turned responsive - widths as % of container.

Duke3D opened the pull request:
I invite you to consider these additions to kwicks.js to make it scale with it's container.

As modified, this js file makes NO changes to the current kwicks behaviour when used with current settings - it only adds a feature triggered by a setting.

The enhancement is to add another setting option - "units" that defaults to 'px' but allows '%'
When in % mode, two other settings have a different meaning - size is maximum container size (max-width) in px and maxSize is maximum image size in px when displayed at 100% (max-width).  
In this mode, all widths and offsets are rendered as %.
(see last commit comment for details for this example.)
example:
container maximum 'size' = 950, image width at maximum 'maxSize' = 300, spacing = .3, units = '%'
rendered styles:
expanded width = 31.7%, unexpanded width (size) = 19.7% for 5 panels, collapsed width = 16.7%
unexpanded layout - 19.7% + .3% + 19.7% + .3% + 19.7% + .3% + 19.7% + .3% + 19.7%  = 99.7%
expanded layout - 16.7% + .3% + 16.7% + .3% + 31.7% + .3% + 16.7% + .3% + 16.7%  = 99.7%

css and html files need to be modified to demo this mode  - would you like that done in the current package / master branch or would you prefer to setup a "responsive" branch or leave this as a separate fork?

Duke

Duke3D committed:
Update jquery.kwicks.js

Add percent as units => fluid

Duke3D committed:
Update jquery.kwicks.js

Type-O Close bracket - Update to handle % as unit and set px as default.

Duke3D committed:
Additional adjustment to manage percentage-based width

Usage:
Optional new configuration setting: units: '%'    (default is 'px')
If selected, these settings have an alternative meaning:
size = size of container in pixels when displayed at 100%.
maxSize = width of image when expanded to display at 100%.
spacing = space between images as a percentage of 'size'.

From this, calculations determine the image size in the traditional sense to be used by the remainder of the function.  

Example for 300x200 images in a container with max-width = 950px:
size: 950,
maxSize: 300,
units: '%',
spacing: 0.3,
...

Result when displayed at full-width (ignoring spacing):
image maximum width = maxWidth (300px) / size (950px) = 31.6%
normal 'size' = container (100%) / no. image panels (6) = 16.4%
image minimum width = (container - maxWidth) / (panels - 1) = 13.3%

As container reduces in size, all image windows reduce proportionately.

Future enhancement - scale images maintaining aspect ratio (reduce height)

jmar777 commented:
Thanks for the PR! Responsive layout is definitely on the roadmap, so this is helpful.  Perhaps we can iterate on the API a little bit though - there are two main features I would like to get in for the next major update:
1. Responsive layout that allows for panel sizes to be a percentage of the container size (as your PR supports), but we need to allow for the following cases:
   - Fixed min size + percentage max size, or
   - Percentage min size + fixed max size, or
   - Both as a percentage, or
   - Both fixed (as it is today)
   - Possibly min/max dimensions for the container itself, but I think we can potentially leave container management external to kwicks - thoughts?
2. Non uniform panel dimensions (i.e., mixed min sizes and max sizes). I see you've also commented on the related [issue](https://github.com/jmar777/kwicks/issues/2) for that.

I anticipate that a clean API for either of these features will require some API breakage (which to me is preferable over a really awkward API), so I'd like to get them in together as a 2.1.0 release, rather than spread it over multiple API-breaking releases.  I'm not entirely sure what the best API for this is though, so I welcome any input on that.

jmar777 commented:
Additional feature that's been requested: programmatically add/remove panels. This probably doesn't need to be supported as part of the same update, but the API should be forwards compatible with it.

Duke3D committed:
Update jquery.kwicks.js

Improve calculations to remove spacing from right edge and solve rounding issues

Duke3D committed:
Update jquery.kwicks.js

Fix type-os causing errors

palafoxernesto commented:
Youre right @r043v  thanks a lot guys!

Mte90 commented:
any news for this pull request?

leigh-johnson commented:
Thank you Duke3D & r043v.

jmar777 commented:
Sorry for the lack of action on this, guys.  @Duke3D, I just pushed a "responsive" branch out to GH (no changes to it yet).  Would it be a massive pain if I asked you to submit the pull request there?  I'll merge it ASAP, I promise!

jmar777 commented:
Strike that. The big green button doesn't allow merging into a different branch, so I just did it manually: https://github.com/jmar777/kwicks/commit/a36f03b449940c2e0fb93992b23468d34342ef86.

Thanks again, @Duke3D!  I'm working on spinning up some examples around it now.  I'll keep this request open until it's been merged back into master.
